File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*; 

class Solution {
    
    static int n, m;
    static boolean visited[][];
    static int ground[][];
    static int[] dx=new int[]{0,0,-1,1};
    static int[] dy=new int[]{-1,1,0,0};
    static HashMap<Integer, Integer> oil = new HashMap<>();

    public int solution(int[][] land) {
        
        n = land.length;
        m = land[0].length;
        visited = new boolean[n][m];
        ground = new int[n][m];
        int index = 1;
        
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(visited[i][j]) continue;
                if(land[i][j] == 0) continue;
                findOil(land, i, j, index);
                index++;
            }
        }
        
        int max = 0;
        HashSet<Integer> get;
        
        for(int i=0; i<m; i++){
            get = new HashSet<>();
            int sum = 0;
            for(int j=0; j<n; j++){
                if(ground[j][i] != 0){
                    get.add(ground[j][i]);
                }
            }
            for(int amount : get){
                sum += oil.get(amount);
            }

            max = Math.max(sum, max);
            
        }
        
        return max;
    }
    
    public void findOil(int[][] land, int x, int y, int index){
        
        Deque<int[]> que = new ArrayDeque<>();
        que.add(new int[] {x, y});
        visited[x][y] = true;
        ground[x][y] = index;
        int sum = 1;
        
        while(!que.isEmpty()){
            
            int now[] = que.poll();
            
            for(int i=0; i<4; i++){
                int nextX = now[0] + dx[i];
                int nextY = now[1] + dy[i];
                if(nextX<0 || nextX>=n || nextY<0 || nextY>=m) continue;
                if(visited[nextX][nextY]) continue;
                if(land[nextX][nextY] == 0) continue;
                que.add(new int[]{nextX, nextY});
                visited[nextX][nextY] = true;
                ground[nextX][nextY] = index;
                sum++;
            }
            
        }
        
        oil.put(index, sum);
        
    }
    
    
} 

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
var n;
var m;
var dx = [0, -1, 0 ,1], dy = [1, 0, -1, 0];
var ground;
var visited;
var map = new Map();

function solution(land) {
    
    n = land.length;
    m = land[0].length;
    ground = Array.from({length : n}, () => Array(m).fill(0));
    visited = Array.from({length : n}, () => Array(m).fill(false));
    var index = 1;
    
    for(let i=0; i<n; i++){
        for(let j=0; j<m; j++){
            if(visited[i][j]) continue;
            if(land[i][j] == 0) continue;
            findOil(land, i, j, index);
            index++;
        }
    }
    
    var max = 0;
    
    for(let i=0; i<m; i++){
        var set = new Set();
        for(let j=0; j<n; j++){
            if(ground[j][i] > 0){
                set.add(ground[j][i]);
            }
        }
        var sum = 0;
        
        set.forEach((id) => sum += map.get(id));
        
        max = Math.max(max, sum);
    }
    
    return max;
}

function findOil(land, x, y, index){
    
    const que = [];
    visited[x][y] = true;
    ground[x][y] = index;
    que.push([x, y]);
    var num = 0;
    var sum = 1;
    
    while(num < que.length){
        var now = que[num];
        for(let i=0; i<4; i++){
            var nextX = now[0] + dx[i];
            var nextY = now[1] + dy[i];
            if(nextX < 0 || nextX >= n || nextY < 0 || nextY >= m) continue;
            if(visited[nextX][nextY]) continue;
            if(land[nextX][nextY] == 0) continue;
            que.push([nextX, nextY]);
            visited[nextX][nextY] = true;
            ground[nextX][nextY] = index;
            sum++
        }
        num++;
    }
    
    map.set(index, sum);
     
}